imgcreate/dnfinst.py

```diff
@@ -195,6 +204,69 @@ def _varSubstitute(option):
         self.repos.add(repo)
         return repo
 
+    def addRepositoryFromConfigFile(self, repo):
+        """
+        Import repository configuration from a DNF repo file instead of
+        from kickstart. If repo is a directory, all *.repo files in that
+        directory are imported.
+        """
+
+        def _to_abs_paths(paths):
+            # replace filenames (relative to repo_dir) with absolute URLs
+            for path in map(str, paths):
+                path = os.path.expanduser(path)
+                absurl = urljoin(
+                    'file://%s/' % os.path.abspath(repo_dir),
+                    path
+                )
+                if not absurl.startswith('file:'):
+                    raise CreatorError(
+                        "livecd-creator does not permit the network URL "
+                        "\"%s\" for GPG keys. Please download this file, "
+                        "manually verify its entire fingerprint, and save "
+                        "it locally." % absurl
+                    )
+                yield absurl
+
+        logging.debug("searching for DNF repositoriesin \"%s\"", repo)
+        if os.path.isdir(repo):
+            repo_dir = repo
+            self.conf.reposdir = (repo)
+            self.conf.config_file_path = '/dev/null'
+        elif os.path.exists(repo):
+            repo_dir = os.path.dirname(repo)
+            self.conf.reposdir = ('/dev/null')
+            self.conf.config_file_path = repo
+        else:
+            raise CreatorError(\
+                "Unable to read repo configuration: \"%s\" does not exist" \
+                    % (repo))
+
+        found_enabled_repos = 0
+        for repo in dnf.conf.read.RepoReader(self.conf, None):
```

Comment:
This class is not explicitly marked as API in dnf… but neither is it explicitly marked as internal. Is it actually available for outside use?

The only alternative I can find is [read_all_repos()](https://dnf.readthedocs.io/en/latest/api_base.html#dnf.Base.read_all_repos). This function makes all errors pass silently; that's definitely not what we want. We also need to modify the configuration prior to inserting the repo.

Reply:
This is probably a question for @j-mracek...

Reply:
Hello @j-mracek, I am looking for an alternative to DNF's [`read_all_repos()`](https://dnf.readthedocs.io/en/latest/api_base.html#dnf.Base.read_all_repos) which throws errors instead of permitting them to pass silently. I'd also like to:

* read configuration from caller-specified files or directories
* inspect/modify the repos before I add them to the `dnf.Base`

Is [`dnf.conf.read.RepoReader`](https://github.com/rpm-software-management/dnf/blob/395541fbf8f87f81cdca7567f22be1182e55bea7/dnf/conf/read.py#L35) considered API? It's not really marked either way.

Reply:
It is correct that [dnf.conf.read.RepoReader](https://github.com/rpm-software-management/dnf/blob/395541fbf8f87f81cdca7567f22be1182e55bea7/dnf/conf/read.py#L35) is not intended as API therefore it is not marked or documented as API. Unfortunately in Python it is quite difficult to hide non API classes or functions.

I am quite curios why [read_all_repos()](https://dnf.readthedocs.io/en/latest/api_base.html#dnf.Base.read_all_repos) is not suitable for you. Disabled repositories in Base().repos are harmless. You can even iterate over enabled repositories using `Base().repos.iter_enabled()`. In case that you wish to modify repository configuration, you can still modify it after loading. What I see there is nothing that you cannot do using standard API. But it requires proper paths in `self.conf.reposdir` and not to replace it with `('/dev/null')`. There is one additional thing that may be you can be interested - dnf.conf can also define repositories.

Reply:
First, from [base.py:547](https://github.com/rpm-software-management/dnf/blob/065c32efb8b1048807b757900e1eff646e699f8f/dnf/base.py#L547)

```py
try:
  self.repos.add(repo)
except dnf.exceptions.ConfigError as e:
  logger.warning(e)
```

I would rather have something which propagates exceptions instead of catching them. While catch is appropriate for a live system, my routine is for image builds. If there's something wrong with the configuration, I want to bail and make the user fix it.

Second, I would like to examine and modify the configuration of each repo I read. My own API that I expose to livecd-creator adds repositories to an existing configuration—it does not clear or replace the existing ones. If I don't somehow track which repos are new, my configuration changes will get applied more than once. For the moment, all the changes I'm making can be done in an idempotent manner, so I guess that's not a huge deal.

In general, I mostly just want to (1) read only the config files that I want to read, and (2) treat warnings as errors, propagating sensible error messages to the caller. `read_all_repos()` can be made to work, but it maybe does more than I'm looking for.

> Unfortunately in Python it is quite difficult to hide non API classes or functions.

The great thing about Python is that every line of code is API. The terrible thing about Python is that *every* line of code is API.

I will refactor this to depend only on that which is explicitly `# :api`.

Reply:
Thanks, it is a safe approach